[yangheng_tang] iP

yanghengtang opened the pull request:
# Chad: Get Your Task Right
Chad is the all-in-one chatbot that helps you manage your day-to-day task and schedule.

## Features
* Different Type of Task: Choose a type of task to **better** fit the reminder you need.
* Ability to Edit task: Add, Delete and Mark Task
* Tasks are Stored Locally: ~No internet connection?~ No worries, the app will still run fine


## Installation Guide
1. Download from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) 
2. Launch the application
3. _You're set to go_ :+1:


## Application Information

### Type of Task
- [ ] `deadline`: A Task consisting of a deadline.
- [ ] `event`: A Task consisting of a start and end date.
- [ ] `todo`: A Task consisting of item to be done.


## User Guide

### Adding Deadlines
To add a Deadline, type `deadline` followed by the description and deadline in `YYYY-MM-DD`.
```
deadline (description) /by: (YYYY-MM-DD)
```

### Adding Events
To add a Event, type `event` followed by the description and start date and end date in `YYYY-MM-DD`.
```
event (description) /start: (YYYY-MM-DD)  /end: (YYYY-MM-DD)
```

### Adding To Do
To add a To Do, type `todo` followed by the description.
```
todo (description)
```

### Marking a Task
To mark a Task, type `mark` followed by the task index.
```
mark (task index)
```

### Delete a Task
To delete a Task, type `mark` followed by the task index.
```
delete (task index)
```

## Fenced Java Out Code
``` java
System.out.println("Hello World!");
```

yanghengtang committed:
Refactor Duke.java

yanghengtang committed:
Add DeadlineException.java

yanghengtang committed:
Add EventException.java

yanghengtang committed:
Add TaskException.java

yanghengtang committed:
Add ToDoException.java

yanghengtang committed:
Update error handling in Duke

yanghengtang committed:
Add duke.txt data file

yanghengtang committed:
Add load file functionality for Duke.java

yanghengtang committed:
Add save file functionality for Duke.java

yanghengtang committed:
Merge branch 'branch-Level-7'

yanghengtang committed:
Update Deadline.java to handle LocalDate type

yanghengtang committed:
Update Event.java to handle LocalDate type

yanghengtang committed:
Update Duke.java to handle LocalDate type

yanghengtang committed:
Update data file

yanghengtang committed:
Merge branch 'branch-Level-8'

yanghengtang committed:
Add Parser.java

yanghengtang committed:
Add Storage.java

yanghengtang committed:
Add TaskList.java

yanghengtang committed:
Add Ui.java

yanghengtang committed:
Refactor Duke.java to work with new classes

yanghengtang committed:
Merge branch 'A-MoreOOP'

yanghengtang committed:
Refactor into packages

yanghengtang committed:
Merge branch 'branch-A-Packages'

yanghengtang committed:
Add Command.java abstract class

yanghengtang committed:
Add AddCommand.java

yanghengtang committed:
Add DeleteCommand.java

yanghengtang committed:
Add ExitCommand.java

yanghengtang committed:
Add MarkCommand.java

yanghengtang committed:
Add PrintListCommand.java

yanghengtang committed:
Add UnknownCommand.java

yanghengtang committed:
Add UnmarkCommand.java

yanghengtang committed:
Refactor Duke.java to include Command class

yanghengtang committed:
Refactor Parser.java to include Command class

yanghengtang committed:
Refactor Ui.java to include Command class

yanghengtang committed:
Merge branch 'add-gradle-support'

yanghengtang committed:
Update data file

yanghengtang committed:
Update mainClass name in build.gradle

yanghengtang committed:
Update DeadlineException.java

yanghengtang committed:
Update EventException.java

yanghengtang committed:
Update TaskException.java

yanghengtang committed:
Update ToDoException.java

yanghengtang committed:
Update Parser.java to handle new Exceptions

yanghengtang committed:
Update TaskList.java to handle out of bound index

yanghengtang committed:
Add ParserTest.java

yanghengtang committed:
Add TaskListTest.java

yanghengtang committed:
Merge branch 'branch-A-JUnit'

yanghengtang committed:
Update build.gradle jar shadowJar archiveFileName

yanghengtang committed:
Add Javadocs for AddCommand.java

yanghengtang committed:
Add Javadocs for Command.java

yanghengtang committed:
Add Javadocs for DeleteCommand.java

yanghengtang committed:
Add Javadocs for ExitCommand.java

yanghengtang committed:
Add Javadocs for MarkCommand.java

yanghengtang committed:
Add Javadocs for PrintListCommand.java

yanghengtang committed:
Add Javadocs for UnknownCommand.java

yanghengtang committed:
Add Javadocs for UnmarkCommand.java

yanghengtang committed:
Add Javadocs for Deadline.java

yanghengtang committed:
Add Javadocs for DeadlineException.java

yanghengtang committed:
Add Javadocs for Event.java

yanghengtang committed:
Add Javadocs for EventException.java

yanghengtang committed:
Add Javadocs for TaskException.java

yanghengtang committed:
Add Javadocs for ToDoException.java

yanghengtang committed:
Add Javadocs for Task.java

yanghengtang committed:
Add Javadocs for ToDo.java

yanghengtang committed:
Add Javadocs for Parser.java

yanghengtang committed:
Add Javadocs for Storage.java

yanghengtang committed:
Add Javadocs for TaskList.java

yanghengtang committed:
Add Javadocs for Ui.java

yanghengtang committed:
Update AddCommand.java code quality

yanghengtang committed:
Update Command.java code quality

yanghengtang committed:
Update DeleteCommand.java code quality

yanghengtang committed:
Update ExitCommand.java code quality

yanghengtang committed:
Update MarkCommand.java code quality

yanghengtang committed:
Update PrintListCommand.java code quality

yanghengtang committed:
Update UnknownCommand.java code quality

yanghengtang committed:
Update UnmarkCommand.java code quality

yanghengtang committed:
Update Deadline.java code quality

yanghengtang committed:
Update Event.java code quality

yanghengtang committed:
Update ToDo.java code quality

yanghengtang committed:
Update Parser.java code quality

yanghengtang committed:
Update Storage.java code quality

yanghengtang committed:
Update Ui.java code quality

yanghengtang committed:
Update Duke.java code quality

yanghengtang committed:
Add support for Find command

yanghengtang committed:
Add FindCommand.java

yanghengtang committed:
Merge branch 'branch-A-JavaDoc'

yanghengtang committed:
Merge branch 'branch-A-CodingStandard'

yanghengtang committed:
Merge branch 'branch-Level-9'

yanghengtang committed:
Fix missing import statement bug

yanghengtang committed:
Add functionality to check and create file or directory if not found

yanghengtang committed:
Update JavaDoc for AddCommand.java

yanghengtang committed:
Update JavaDoc for Command.java

yanghengtang committed:
Update JavaDoc for DeleteCommand.java

yanghengtang committed:
Update JavaDoc for ExitCommand.java

yanghengtang committed:
Add JavaDoc for FindCommand.java

yanghengtang committed:
Update JavaDoc for MarkCommand.java

yanghengtang committed:
Update JavaDoc for PrintListCommand.java

yanghengtang committed:
Update JavaDoc for UnknownCommand.java

yanghengtang committed:
Update JavaDoc for UnmarkCommand.java

yanghengtang committed:
Update JavaDoc for Parser.java

yanghengtang committed:
Refactor Parser and Storage for parsing of TaskList to data

yanghengtang committed:
Update Ui.java logic

yanghengtang committed:
Refactor to work with updated Ui logic

yanghengtang committed:
Update error handling for FileNotFoundException

yanghengtang committed:
Add checkstyle.xml

yanghengtang committed:
Add suppressions.xml

yanghengtang committed:
Update build.gradle to include Checkstyle

yanghengtang committed:
Fix Duke.java style

yanghengtang committed:
Fix AddCommand.java style

yanghengtang committed:
Fix DeleteCommand.java style

yanghengtang committed:
Fix FindCommand.java style

yanghengtang committed:
Fix MarkCommand.java style

yanghengtang committed:
Fix UnmarkCommand.java style

yanghengtang committed:
Fix Deadline.java style

yanghengtang committed:
Fix Event.java style

yanghengtang committed:
Fix Task.java style

yanghengtang committed:
Fix ToDoException.java style

yanghengtang committed:
Fix Parser.java style

yanghengtang committed:
Fix Storage.java style

yanghengtang committed:
Fix TaskList.java style

yanghengtang committed:
Fix Ui.java style

yanghengtang committed:
Merge branch 'branch-A-CheckStyle'

yanghengtang committed:
Add DialogBox.java

yanghengtang committed:
Add Launcher.java

yanghengtang committed:
Add Main.java

yanghengtang committed:
Add MainWindow.java

yanghengtang committed:
Add JavaFX dependencies into build.gradle

yanghengtang committed:
Update Duke.java with getResponse method

yanghengtang committed:
Add resources for GUI

yanghengtang committed:
Update getResponse method logic

yanghengtang committed:
Update Ui.java to work with GUI

yanghengtang committed:
Update Chad image

yanghengtang committed:
Add new Chad photo

yanghengtang committed:
Update DialogBox Style

yanghengtang committed:
Merge branch 'branch-Level-10'

ph-nathan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,56 @@
+        } finally {
+            output = ui.showMessage();
+        }
+        return output;
```
woah very clean code, encapsulation over 9000.

ph-nathan commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Prints the current message in the Ui object and clears the message.
+     */
+    public void printMessage() {
```
Maybe can consider removing this method just for cleaner codebase since you're not using it anymore :) (lol could finally nitpick something)

ph-nathan commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Prints the entry message of the chatbot.
+     */
+    public String printEntryMessage() {
```
I think can either 
utilize this method to set the welcome message of your GUI since you're returning a string here! (can consider making this method static)
or if not can remove for cleanliness

ph-nathan commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,52 @@
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
```
```suggestion
        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
        dialogContainer.getChildren().add(DialogBox.getDukeDialog(Ui.printEntryMessage(), dukeImage));
```

ph-nathan replied:
here's a way you can add the entry message on program's launch!

ph-nathan commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+public abstract class Command {
+    private boolean exit;
+
+    public Command(boolean exit) {
```
I think since Command is an abstract class, we can't instantiate it, so a constructor here is kind of not needed? Do correct me if I'm missing something though!

ph-nathan reviewed (commented):
Overall, the codebase was really clean, and the encapsulation aspect was well-achieved. I only suggested some possible improvements on features and nitpicked a few small details to further clean out the code. No major coding standard violation or naming problem. Good job bro, I took note of some stuff for myself to improve just from reviewing your code!

ruo-x commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
+        TaskList filteredTaskList = taskList.find(this.queryString);
```
I like how you named the methods you use, very easy to understand!

ruo-x commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,38 @@
+     * @param taskList The given TaskList with the Task to be unmarked.
+     * @param ui The given Ui to show the status of the unmarking of the task.
+     * @param storage The given Storage that saves the TaskList locally.
+     */
```
Perhaps you can consider combining all the commands into a class since they only consist of one method?

ruo-x commented on src/main/java/duke/task/DeadlineException.java:
```diff
@@ -0,0 +1,12 @@
+public class DeadlineException extends TaskException {
+
+    public DeadlineException() {
+        super("Valid Format: deadline (description) /by (YYYY-MM-DD)\n");
```
I like how the correct input format is provided for the users.👍

ruo-x commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Represents a task with status indicating its completeness
+ */
+public class Task {
```
Maybe you can consider making Task an abstract class since it should not be instantiated?

ruo-x commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,121 @@
+     */
+    public Task remove(int index) {
+        if (index < 0 || index >= list.size()) {
+            return null;
```
Perhaps you can consider throwing an exception when the user has keyed in an invalid index, just so they know it was unsuccessful?

ruo-x reviewed (commented):
Overall, your code is very clean and easy to understand! Added in a few of my suggestions that you can consider :)

yanghengtang committed:
Add Assertions for Deadline.java

yanghengtang committed:
Add Assertions for Event.java

yanghengtang committed:
Add Assertions for Task.java

yanghengtang committed:
Add Assertions for Parser.java

yanghengtang committed:
Add Assertions for Storage.java

yanghengtang committed:
Add Assertions for TaskList.java

yanghengtang committed:
Add Assertions for Ui.java

yanghengtang committed:
Fix Style for DialogBox.java

yanghengtang committed:
Add JavaDoc for Launcher.java

yanghengtang committed:
Add JavaDoc for Ui.java

yanghengtang committed:
Update Duke.java Code Quality

yanghengtang committed:
Add toDataFormatString method for Deadline.java

yanghengtang committed:
Add toDataFormatString method for Event.java

yanghengtang committed:
Update Task.java to be abstract and add toDataFormatString abstract method

yanghengtang committed:
Add toDataFormatString method for ToDo.java

yanghengtang committed:
Refactor Parser.java parseTaskListToData method

yanghengtang committed:
Add get method for TaskList.java

yanghengtang committed:
Update AddCommand.java code quality

yanghengtang committed:
Update Deadline.java code quality

yanghengtang committed:
Update Event.java code quality

yanghengtang committed:
Update Task.java code quality

yanghengtang committed:
Update ToDo.java code quality

yanghengtang committed:
Update Parser.java code quality

yanghengtang committed:
Update TaskList.java code quality

yanghengtang committed:
Refactor DialogBox.java to gui package

yanghengtang committed:
Refactor Launcher.java to gui package

yanghengtang committed:
Refactor Main.java to gui package

yanghengtang committed:
Refactor MainWindow.java to gui package

yanghengtang committed:
Update build.gradle application mainClass

yanghengtang committed:
Merge pull request #1 from yanghengtang/branch-A-Assertions

Merge branch A-Assertions

yanghengtang committed:
Merge pull request #2 from yanghengtang/branch-A-CodeQuality

Merge branch A-CodeQuality

yanghengtang committed:
Add GitHub Actions to set up Continuous Integration

yanghengtang committed:
Fix Parse.java style

yanghengtang committed:
Remove text-ui-test

yanghengtang committed:
Fix ParserTest.java style

yanghengtang committed:
Fix TaskListTest.java style

yanghengtang committed:
Update GitHub CI gradle.yml

yanghengtang committed:
Update DateTime Format of the dates displayed

The deliverable of Level-8 requires the Datee object to be displayed
in a different format as the standard string representation.

Utilise format method of LocalDate class and use ofPattern method from
DateTimeFormatter class to specify the desired format.

yanghengtang committed:
Update Deadline.java to handle LocalDateTime

yanghengtang committed:
Update Event.java to handle LocalDateTime

yanghengtang committed:
Update Parser.java to handle LocalDateTime

yanghengtang committed:
Update data file to handle LocalDateTime

yanghengtang committed:
Add Comparable interface for Deadline.java

yanghengtang committed:
Add Comparable interface for Event.java

yanghengtang committed:
Add Comparable interface for ToDo.java

yanghengtang committed:
Add Comparable interface for Task.java

yanghengtang committed:
Add TaskComparator for List sorting

yanghengtang committed:
Fix bug in Parser.java

yanghengtang committed:
Add sorting functionality to TaskList

yanghengtang committed:
Fix test case to work with LocalDateTime

yanghengtang committed:
Update local data file

yanghengtang committed:
Merge branch 'branch-C-Sort'

yanghengtang committed:
Update DialogBox GUI

yanghengtang committed:
Update Code Quality for Command.java

yanghengtang committed:
Update Code Quality for Storage.java

yanghengtang committed:
Update Code Quality for TaskComparator.java

yanghengtang committed:
Update Code Quality for TaskList.java

yanghengtang committed:
Update Code Quality for Ui.java

yanghengtang committed:
Update README.md

yanghengtang committed:
Rename DialogBox to UserDialogBox

yanghengtang committed:
Add ChadDialogBox.java

yanghengtang committed:
Add ChadDialogBox.fxml

yanghengtang committed:
Update MainWindow with ChadDialogBox

yanghengtang committed:
Update UserDialogBox

yanghengtang committed:
Add Welcome Message in MainWindow

yanghengtang committed:
Update JavaDoc for MainWindow

yanghengtang committed:
Update welcome message for Ui

yanghengtang committed:
Update ChadDialogBox GUI

yanghengtang committed:
Update UserDialogBox GUI

yanghengtang committed:
Update MainWindow GUI

yanghengtang committed:
Merge branch 'branch-A-BetterGui'

yanghengtang committed:
Package exception into duke.exception

yanghengtang committed:
Add DeleteException.java

yanghengtang committed:
Add DukeException.java

yanghengtang committed:
Add FindException.java

yanghengtang committed:
Add LoadException.java

yanghengtang committed:
Add MarkException.java

yanghengtang committed:
Add UnmarkException.java

yanghengtang committed:
Update CommandException.java

yanghengtang committed:
Update DeadlineException.java

yanghengtang committed:
Update EventException.java

yanghengtang committed:
Update ToDoException.java

yanghengtang committed:
Update classes to handle new Exceptions

yanghengtang committed:
Merge branch 'branch-A-MoreErrorHandling'

yanghengtang committed:
Update ParserTest to handle new Exception

yanghengtang committed:
Update ParserTest style

yanghengtang committed:
Update README.md

yanghengtang committed:
Merge branch 'branch-A-UserGuide'

yanghengtang committed:
Update README.md title

yanghengtang committed:
Refactor Duke to Chad

yanghengtang committed:
Update README.md

yanghengtang committed:
Refactor DukeException into ChadException

yanghengtang committed:
Refactor duke package into chad

yanghengtang committed:
Add Save feature to other commands

yanghengtang committed:
Update build.gradle mainClass

yanghengtang committed:
Update Chad to handle new Exceptions thrown

yanghengtang committed:
Update MainWindow to close upon ExitCommand

yanghengtang committed:
Refactor parseFile method from Storage to Parser

yanghengtang committed:
Update Code Quality for TaskList

yanghengtang committed:
Update Code Quality for Ui.java

yanghengtang committed:
Update local data file

yanghengtang committed:
Add Product screenshot

yanghengtang committed:
Remove isExit from Command classes

yanghengtang committed:
Update product screenshot

yanghengtang committed:
Add setTitle for GUI

yanghengtang committed:
Update local data file

yanghengtang committed:
Add SaveException.java

yanghengtang committed:
Update Command classes to handle SaveException

yanghengtang committed:
Update style for SaveException

yanghengtang committed:
Remove DaDuke.png

yanghengtang committed:
Update Code Quality

yanghengtang committed:
Update README.md

yanghengtang committed:
Add task list size message for PrintList Command

yanghengtang committed:
Update build.gradle

yanghengtang committed:
Update build.gradle

yanghengtang committed:
Update error handling for Mark, Unmark and Delete Command
